adage/analyze/api.py

```diff
@@ -212,6 +213,36 @@ class Meta:
         queryset = Node.objects.all()
         resource_name = 'node'
         allowed_methods = ['get']
+        filtering = {
+            'name': ('exact', 'in', ),
+            'heavy_genes': ('exact', ),  # New filter, see apply_filters().
+        }
+
+    def apply_filters(self, request, applicable_filters):
+        object_list = super(NodeResource, self).apply_filters(
+            request, applicable_filters)
+        heavy_genes = request.GET.get('heavy_genes', None)
+        if heavy_genes:
+            # Instead of relying on tastypie/resources.py to catch the
+            # ValueError exception that may be raised in this function,
+            # we catch the one that may be raised by int() below and
+            # raise a customized BadRequest exception with more details.
+            try:
+                # Convert genes to a set of integers so that the
+                # duplicate(s) will be removed implicitly.
+                query_genes = {int(id) for id in heavy_genes.split(',')}
+            except ValueError:
+                raise BadRequest("Invalid gene IDs: %s" % heavy_genes)
+
+            for (i, q) in enumerate(query_genes):
+                q_nodes = {p.node.id for p in
+                           Participation.objects.filter(gene=q)}
+                if i == 0:
+                    related_nodes = q_nodes
```

Comment:
Since the **for** loop does not have its own scope (one of the features in C++/Java that I miss most when writing python), it doesn't really matter whether **related_nodes** is initialized inside or outside the loop.